pkg/auth/github.go

```diff
@@ -0,0 +1,163 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+// codeResponse represents response body from the Github API
+type codeResponse struct {
+	DeviceCode     string `json:"device_code"`
+	UserCode       string `json:"user_code"`
+	VerficationURI string `json:"verification_uri"`
+	Interval       int    `json:"interval"`
+}
+
+// Uniquely identifies us as a GitHub app.
+// This does not need to be obfuscated because it is publicly available
+// to anyone who does an OAuth request via wego.
+// See the auth ADR for more details:
+// https://github.com/weaveworks/weave-gitops/blob/main/doc/adr/0005-wego-core-auth-strategy.md#design
+const WeGOGithubClientID = "edcb13588d46f254052c"
+
+// Encapsulate shared logic between doCodeRequest and doAuthRequest
+func doRequest(req *http.Request, client *http.Client) ([]byte, error) {
+	req.Header.Set("Accept", "application/json")
+
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		// err is falsey even on 4XX or 5XX
+		return nil, err
+	}
+
+	rb, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+// doCodeRequest does the initial request to of the Device Flow
+func doCodeRequest(client *http.Client, scope string) (*codeResponse, error) {
+	deviceCodeQuery := fmt.Sprintf("client_id=%s&scope=%s", WeGOGithubClientID, scope)
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("https://github.com/login/device/code?%s", deviceCodeQuery), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := doRequest(req, client)
+	if err != nil {
+		return nil, fmt.Errorf("error doing code request: %w", err)
+	}
+
+	d := &codeResponse{}
+
+	if err := json.Unmarshal(b, d); err != nil {
+		return nil, fmt.Errorf("could not unmarshal code response: %w", err)
+	}
+
+	return d, nil
+}
+
+var ErrAuthPending = errors.New("auth pending")
+
+const accessTokenUrlTemplate = "https://github.com/login/oauth/access_token?client_id=%s&device_code=%s&grant_type=%s"
+const githubRequredGrantType = "urn:ietf:params:oauth:grant-type:device_code"
+
+// It appears we need `repo` scope, which is VERY permissive.
+// We need to be able to push a deploy key and merge commits. No other scopes matched.
+// Available scopes: https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps
+const githubOAuthScope = "repo"
+
+type githubAuthResponse struct {
+	AccessToken string `json:"access_token"`
+	Error       string `json:"error"`
+}
+
+// doAuthRequest is used to poll for the status of the device flow.
+func doAuthRequest(client *http.Client, deviceCode string) (string, error) {
+	url := fmt.Sprintf(accessTokenUrlTemplate, WeGOGithubClientID, deviceCode, githubRequredGrantType)
```

Comment:
FYI: url.URL is an alternative way to do this and a lot cleaner for more complex URLs.

Reply:
I looked around for some sort of way to build a query from `map[string]string`. Is there a builtin for that?

Reply:
Looks like this might work: https://pkg.go.dev/net/url#Values.Encode

Reply:
Refactored to use the `Values.Encode()` function